src/Decoder/EntryTransactionDetail.php

```diff
@@ -69,50 +71,49 @@ public function addRelatedParties(DTO\EntryTransactionDetail $detail, SimpleXMLE
+                $this->addRelatedParty($detail, $xmlRelatedPartyType, DTO\Creditor::class, $xmlRelatedPartyTypeAccount);
             }
 
             if (isset($xmlRelatedParty->UltmtCdtr)) {
                 $xmlRelatedPartyType = $xmlRelatedParty->UltmtCdtr;
-                $xmlRelatedPartyName = (isset($xmlRelatedPartyType->Nm)) ? (string) $xmlRelatedPartyType->Nm : '';
-                $relatedPartyType = new DTO\UltimateCreditor($xmlRelatedPartyName);
 
-                $this->addRelatedParty($detail, $xmlRelatedPartyType, $relatedPartyType);
+                $this->addRelatedParty($detail, $xmlRelatedPartyType, DTO\UltimateCreditor::class);
             }
 
             if (isset($xmlRelatedParty->Dbtr)) {
                 $xmlRelatedPartyType = $xmlRelatedParty->Dbtr;
                 $xmlRelatedPartyTypeAccount = $xmlRelatedParty->DbtrAcct;
-                $xmlRelatedPartyName = (isset($xmlRelatedPartyType->Nm)) ? (string) $xmlRelatedPartyType->Nm : '';
-                $relatedPartyType = $debtor = new DTO\Debtor($xmlRelatedPartyName);
 
-                $this->addRelatedParty($detail, $xmlRelatedPartyType, $relatedPartyType, $xmlRelatedPartyTypeAccount);
+                $this->addRelatedParty($detail, $xmlRelatedPartyType, DTO\Debtor::class, $xmlRelatedPartyTypeAccount);
             }
 
             if (isset($xmlRelatedParty->UltmtDbtr)) {
                 $xmlRelatedPartyType = $xmlRelatedParty->UltmtDbtr;
-                $xmlRelatedPartyName = (isset($xmlRelatedPartyType->Nm)) ? (string) $xmlRelatedPartyType->Nm : '';
-                $relatedPartyType = new DTO\UltimateDebtor($xmlRelatedPartyName);
 
-                $this->addRelatedParty($detail, $xmlRelatedPartyType, $relatedPartyType);
+                $this->addRelatedParty($detail, $xmlRelatedPartyType, DTO\UltimateDebtor::class);
             }
         }
     }
 
-    protected function addRelatedParty(DTO\EntryTransactionDetail $detail, SimpleXMLElement $xmlRelatedPartyType, DTO\RelatedPartyTypeInterface $relatedPartyType, ?SimpleXMLElement $xmlRelatedPartyTypeAccount = null): RelatedParty
+    /**
+     * @param class-string<RelatedPartyTypeInterface> $relatedPartyTypeClass
+     */
+    protected function addRelatedParty(DTO\EntryTransactionDetail $detail, SimpleXMLElement $xmlRelatedPartyType, string $relatedPartyTypeClass, ?SimpleXMLElement $xmlRelatedPartyTypeAccount = null): void
     {
-        if (isset($xmlRelatedPartyType->PstlAdr)) {
-            $relatedPartyType->setAddress(DTOFactory\Address::createFromXml($xmlRelatedPartyType->PstlAdr));
+        // CAMT v08 uses substructure, so we check for its existence or fallback to the element itself to keep compatibility with CAMT v04
+        $xmlPartyDetail = $xmlRelatedPartyType->Pty ?: $xmlRelatedPartyType->Agt?->FinInstnId ?: $xmlRelatedPartyType;
+
+        $xmlRelatedPartyName = (isset($xmlPartyDetail->Nm)) ? (string) $xmlPartyDetail->Nm : null;
```

Comment:
Also, `FinInstnId `is never referencing `Nm`. In fact I don't find any specification for `FinInstnId`. Fo me it is only an identifier. Where have you found the structrure for your tests?

Reply:
[`FinInstnId` may have a `Nm`](https://github.com/genkgo/camt/blob/211c7579c6679245b9045c132feea86820123c80/assets/camt.054.001.08.xsd#L862), as you can [see in my tests](https://github.com/genkgo/camt/blob/211c7579c6679245b9045c132feea86820123c80/test/data/camt054.v8-with-financial-institution.xml#L31) which pass XSD validation. And it will be [properly read into our (simplified) model](https://github.com/genkgo/camt/blob/211c7579c6679245b9045c132feea86820123c80/test/data/camt054.v8-with-financial-institution.json#L87).

Reply:
@PowerKiKi you are right, the XSD is defining FinInstnId as you described. Thank you for the reference. Weird that the documentation doesn’t explain this structure.